### PR TITLE
drivers: spi: spi_ll_stm32: Add SPI support for STM32MP1x SoC and stm32mp157c_dk2 board

### PR DESCRIPTION
**Pull request by yboutreux**
Abstract some SPI LL function call for future driver compatibility with a new SPI peripheral version (introduced with STM32MP1x and STM32H7x SoC).
Add SPI support for STM32MP1x SoC.
Add SPI support for stm32mp157c_dk2 board. If SPI is selected, SPI4 (Arduino connector compatible SPI) and SPI5 (on front 2x20 GPIO expander) will be enable by default on stm32mp157c_dk2 board.

More specific informations about "drivers: spi: spi_ll_stm32: set NSS before mode" commit : 
stm32mp1x (and stm32h7) LL function _SPI_Init_ seems to also define first the NSS then the mode, unlike other STM32 boards where this is not specified.
Changing the order shouldn't have bad repercussions on other boards, Zephyr SPI driver test has been passed successfully on disco_l475_iot1 board

### CODE REVIEW

File: drivers/spi/spi_ll_stm32.c [hunk 1]
```diff
@@ -318,6 +312,12 @@ static int spi_stm32_configure(struct device *dev,
 		}
 	}
 
+	if (config->operation & SPI_OP_MODE_SLAVE) {
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: If MP1 specific, are we sure this is safe for other series?
If yes, please state it in commit message.

File: drivers/spi/spi_ll_stm32.c [hunk 2]
```diff
@@ -41,6 +48,8 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
 #define SPI_STM32_ERR_MSK (LL_SPI_SR_CRCERR | LL_SPI_SR_MODF | LL_SPI_SR_OVR)
 #endif
 
+#endif /* STM32MP1 */
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: CONFIG_SOC_SERIES_STM32MP1X

File: drivers/spi/spi_ll_stm32.c [hunk 3]
```diff
@@ -31,6 +31,13 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
  * error flag, because STM32F1 SoCs do not support it and  STM32CUBE
  * for F1 family defines an unused LL_SPI_SR_FRE.
  */
+
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+#define SPI_STM32_ERR_MSK (LL_SPI_SR_UDR | LL_SPI_SR_CRCE | LL_SPI_SR_MODF | \
+			   LL_SPI_SR_OVR | LL_SPI_SR_TIFRE)
+
+#else /* STM32MP1 */
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: CONFIG_SOC_SERIES_STM32MP1X

File: drivers/spi/spi_ll_stm32.c [hunk 4]
```diff
@@ -90,9 +99,26 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
...
+	/* STM32MP1 has some register name modifications :
+	 * LL_SPI_IsActiveFlag_TXP <-> LL_SPI_IsActiveFlag_TXE
+	 * LL_SPI_IsActiveFlag_RXP <-> LL_SPI_IsActiveFlag_RXNE
+	 * Also, if the device is the SPI master, we need to enable the start of
+	 * the transfer with LL_SPI_StartMasterTransfer(spi)
+	 */
+	while (!LL_SPI_IsActiveFlag_TXP(spi)) {
+		/* NOP if not enough space for a packet*/
+	}
+	if (LL_SPI_IsEnabled(spi) && LL_SPI_GetMode(spi)) {
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: `LL_SPI_GetMode(spi)` likely doesn't return true/false, please test against `LL_SPI_MODE_MASTER`.
OTOH, I think `LL_SPI_IsEnabled` will logically return true here.

File: drivers/spi/spi_ll_stm32.c [hunk 5]
```diff
@@ -90,9 +99,26 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 	u16_t rx_frame;
 
 	tx_frame = spi_stm32_next_tx(data);
+
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+	/* STM32MP1 has some register name modifications :
+	 * LL_SPI_IsActiveFlag_TXP <-> LL_SPI_IsActiveFlag_TXE
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: In that case, might be good to add a level of abstraction here instead of ifdefing the whole code, cd what was done here:
https://github.com/zephyrproject-rtos/zephyr/tree/master/drivers/ipm/ipm_stm32_ipcc.c

File: drivers/spi/spi_ll_stm32.c [hunk 6]
```diff
@@ -103,9 +129,18 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 		spi_context_update_tx(&data->ctx, 2, 1);
 	}
 
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+	while (!LL_SPI_IsActiveFlag_RXP(spi)) {
+		/* NOP if buffer has less than a packet to read*/
+	}
+	if (LL_SPI_IsEnabled(spi) && LL_SPI_GetMode(spi)) {
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: ditto

File: drivers/spi/spi_ll_stm32.c [hunk 7]
```diff
@@ -178,22 +244,48 @@ static void spi_stm32_complete(struct spi_stm32_data *data, SPI_TypeDef *spi,
...
 		(void) LL_SPI_ReceiveData8(spi);
 	}
-#endif
+#endif	/* CONFIG_SOC_SERIES_STM32MP1X */
+#endif	/* CONFIG_SPI_STM32_HAS_FIFO */
 
 	if (LL_SPI_GetMode(spi) == LL_SPI_MODE_MASTER) {
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+		while (LL_SPI_IsEnabled(spi) &&
+		       !LL_SPI_IsActiveFlag_MODF(spi) &&
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: Not sure about these conditione, if we check what is done in the while block

File: west.yml [hunk 8]
```diff
@@ -51,7 +51,7 @@ manifest:
       revision: 0ec40aed8087f26bd9ac1b70fb5a6c326a6451aa
       path: modules/hal/st
     - name: hal_stm32
-      revision: 686c93da4186eb9c48b83ee235fc0454693f6e5c
```
Comment (erwango) [inline west.yml]: Split in dedicated commit

File: dts/arm/st/mp1/stm32mp157.dtsi [hunk 9]
```diff
@@ -147,6 +147,56 @@
 			};
 		};
 
+		spi1: spi@44004000 {
```
Comment (erwango) [inline dts/arm/st/mp1/stm32mp157.dtsi]: Split in dedicated commit, along with fixup changes

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by erwango** [comment]
DNM for now (conflicts with STM32 I2C change in hal-stm32)

File: drivers/spi/spi_ll_stm32.c (hunk 1 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: stm32mp1x (and stm32h7) LL function SPI_Init seems to also define first the NSS then the mode, unlike other STM32 boards where this is not specified.
Changing the order shouldn't have bad repercussions on other boards, Zephyr SPI driver test has been passed successfully on disco_l475_iot1 board

File: drivers/spi/spi_ll_stm32.c (hunk 4 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: `LL_SPI_IsEnabled` return either 1 (true) or 0 (false)
`LL_SPI_GetMode(spi)` return either :
```
#define LL_SPI_MODE_MASTER                      (SPI_CFG2_MASTER)
#define LL_SPI_MODE_SLAVE                          (0x00000000UL)
```

I agree this might not be super pretty to rely on the slave to be 0, and the master >0, do you prefer : 
`if (LL_SPI_IsEnabled(spi) && (LL_SPI_GetMode(spi) ==  LL_SPI_MODE_MASTER)) {`

File: drivers/spi/spi_ll_stm32.c (hunk 6 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: Same response :
`LL_SPI_IsEnabled` return either 1 (true) or 0 (false)
`LL_SPI_GetMode(spi)` return either :
```
#define LL_SPI_MODE_MASTER                      (SPI_CFG2_MASTER)
#define LL_SPI_MODE_SLAVE                          (0x00000000UL)
```

I agree this might not be super pretty to rely on the slave to be 0, and the master >0, do you prefer : 
`if (LL_SPI_IsEnabled(spi) && (LL_SPI_GetMode(spi) ==  LL_SPI_MODE_MASTER)) {`

File: drivers/spi/spi_ll_stm32.c (hunk 7 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: This is just an adaptation to the BSY flag which isn't existing on STM32MP1 SPI peripheral registers.
Definition taken from _stm32l4xx_ll_spi.h_ : 
```
/**
  * @brief  Get busy flag
  * @note   The BSY flag is cleared under any one of the following conditions:
  * -When the SPI is correctly disabled
  * -When a fault is detected in Master mode (MODF bit set to 1)
  * -In Master mode, when it finishes a data transmission and no new data is ready to be
  * sent
  * -In Slave mode, when the BSY flag is set to '0' for at least one SPI clock cycle between
  * each data transfer.
  * @rmtoll SR           BSY           LL_SPI_IsActiveFlag_BSY
  * @param  SPIx SPI Instance
  * @retval State of bit (1 or 0).
  */
```
Regarding the while loop content, this is something that is introduced here https://github.com/zephyrproject-rtos/zephyr/pull/17804

File: dts/arm/st/mp1/stm32mp157.dtsi (hunk 9 above)
Comment (yboutreux) [reply dts/arm/st/mp1/stm32mp157.dtsi]: Splitted, I also added changes related to STM32MP1x SoC, so commit are split in soc support, driver support and board support

File: drivers/spi/spi_ll_stm32.c (hunk 4 above)
Comment (erwango) [reply drivers/spi/spi_ll_stm32.c]: > LL_SPI_IsEnabled return either 1 (true) or 0 (false)

Well, at this step, should always be 1, isn't it?

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks.

File: drivers/spi/spi_ll_stm32.c (hunk 5 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: That's a good idea, I'll do that ! 
The equivalence is not straight forward, but I think the behavior should be the same  :

**Other MCUs : TXE**
0: Tx buffer not empty
1: Tx buffer empty
**MP1 : TXP**
0: there is not enough space to locate next data packet at TxFIFO
1: TxFIFO has enough free location to host 1 data packet
_________________________________________

**Other MCUs : RXNE**
0: Rx buffer empty
1: Rx buffer not empty
**MP1 : RXP**
0: RxFIFO is empty or a not complete data packet is received
1: RxFIFO contains at least 1 data packet

File: drivers/spi/spi_ll_stm32.c (hunk 4 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: It should, I had odd behaviors and errors with the driver, and I had to check if the SPI is still enabled or if we encountered an issue that disabled the SPI.
Now this is fixed, and we don't need to check if the SPI is enabled, so this part of the condition has been removed

**Comment by yboutreux** [comment]
Update : add abstraction to some LL function, since the adding of a new SPI peripheral overburden the code with some specific cases related to STM32MP1x SoC (and soon STM32H7x SoC which has the same SPI peripheral version)

**Comment by yboutreux** [comment]
Rebased to remove merge conflicts

File: west.yml (hunk 8 above)
Comment (yboutreux) [reply west.yml]: Missed this comment, this is done thank you !

**Review by arnopo — APPROVED** [review]

File: drivers/spi/spi_ll_stm32.c [hunk 10]
```diff
@@ -307,12 +307,6 @@ static int spi_stm32_configure(struct device *dev,
 
 	LL_SPI_DisableCRC(spi);
```
Comment (avisconti) [inline drivers/spi/spi_ll_stm32.c]: "This commit solves an issue where the NSS must be set before the
mode on the stm32mp157c_dk2, else LL_SPI_SetMode won't affect the
mode registry."

Do you mean that there is an h/w issue you are taking care of?
Does it make sense to mention some offiicial documentation about it?

Comment (arnop2) [reply drivers/spi/spi_ll_stm32.c]: @avisconti 
No it is a SW bug. The SPI configuration sequence is not respected (please refer to stm32MP1 ref manual for details). If NSS is configured after the mode is set then a MODF error can occur.

File: drivers/spi/spi_ll_stm32.h [hunk 11]
```diff
@@ -25,57 +25,126 @@ struct spi_stm32_data {
...
 }
 
 static inline void spi_stm32_ll_func_DisableIT_ERR(SPI_TypeDef *spi)
 {
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+	LL_SPI_DisableIT_UDR(spi);
+	LL_SPI_DisableIT_OVR(spi);
+	LL_SPI_DisableIT_CRCERR(spi);
+	LL_SPI_DisableIT_FRE(spi);
+	LL_SPI_DisableIT_MODF(spi);
```
Comment (avisconti) [inline drivers/spi/spi_ll_stm32.h]: I see all this changes for MP1. But is it really necessary to do it at zephyr level?
Shouldn't be a better idea to incorporate the SoC difference in LL driver?
In this case we will always call LL_SPI_DisableIT_ERR() which will do the proper thing for MP1.

File: soc/arm/st_stm32/stm32mp1/soc.h [hunk 12]
```diff
@@ -46,6 +46,10 @@
 #include <stm32mp1xx_ll_usart.h>
 #endif
 
+#ifdef CONFIG_SPI_STM32
```
Comment (erwango) [inline soc/arm/st_stm32/stm32mp1/soc.h]: This change should be with commit where you implement MP1 support in SPI driver

File: drivers/spi/spi_ll_stm32.c [hunk 13]
```diff
@@ -90,7 +90,7 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 	u16_t rx_frame;
 
 	tx_frame = spi_stm32_next_tx(data);
-	while (!LL_SPI_IsActiveFlag_TXE(spi)) {
+	while (!spi_stm32_ll_func_IsActiveFlag_TXE(spi)) {
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: New functions, which are zephyr specific should follow zephyr coding style, hence should be lower_case (not CamelCase)

File: drivers/spi/spi_ll_stm32.c [hunk 14]
```diff
@@ -178,22 +179,22 @@ static void spi_stm32_complete(struct spi_stm32_data *data, SPI_TypeDef *spi,
 			       int status)
 {
 #ifdef CONFIG_SPI_STM32_INTERRUPT
-	LL_SPI_DisableIT_TXE(spi);
-	LL_SPI_DisableIT_RXNE(spi);
-	LL_SPI_DisableIT_ERR(spi);
+	spi_stm32_ll_func_DisableIT_TXE(spi);
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: Can we group these under `spi_stm32_ll_func_disable_it` ?

File: drivers/spi/spi_ll_stm32.h [hunk 15]
```diff
@@ -23,4 +23,59 @@ struct spi_stm32_data {
 	struct spi_context ctx;
 };
 
+static inline u32_t spi_stm32_ll_func_IsActiveFlag_TXE(SPI_TypeDef *spi)
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.h]: Are these function names sufficiently abstracted?
Also I think you can spare the `spi_stm32` to minimize function names length.

File: drivers/spi/spi_ll_stm32.h (hunk 11 above)
Comment (erwango) [reply drivers/spi/spi_ll_stm32.h]: @avisconti, this is a separate topic that should be discussed internally.

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/stm32mp157c_dk2/Kconfig.defconfig [hunk 16]
```diff
@@ -24,4 +24,17 @@ config UART_7
...
 
+if SPI
+
+config SPI_STM32_INTERRUPT
+	default y
+
+config SPI_4
+	default y
+
+config SPI_5
```
[the remark this answers is not in the document]
Comment (yboutreux) [reply boards/arm/stm32mp157c_dk2/Kconfig.defconfig]: No you are right, is we enable SPI then all SPI should be available, thank you ! :)

File: drivers/spi/spi_ll_stm32.c (hunk 14 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: Since we need to keep ll_func_enable_IT_TXE(spi), ll_func_enable_IT_RXNE(spi), ll_func_enable_IT_ERR(spi), ll_func_disable_IT_TXE(spi), ll_func_disable_IT_RXNE(spi), I thought that keeping ll_func_disable_IT_ERR(spi) instead of grouping those three lines was better for consistency sake

File: drivers/spi/spi_ll_stm32.h (hunk 15 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.h]: I see what you mean, we could take the opportunity to do "is_rx_empty" "is_tx_not_empty" "is_busy"... On my way !

File: drivers/spi/spi_ll_stm32.h [hunk 17]
```diff
@@ -23,4 +23,59 @@ struct spi_stm32_data {
 	struct spi_context ctx;
 };
 
+static inline u32_t ll_func_TX_is_empty(SPI_TypeDef *spi)
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.h]: Not fully coding style compliant yet...

File: drivers/spi/spi_ll_stm32.c [hunk 18]
```diff
@@ -31,6 +31,13 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
  * error flag, because STM32F1 SoCs do not support it and  STM32CUBE
  * for F1 family defines an unused LL_SPI_SR_FRE.
  */
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+
+#define LL_SPI_SR_CRCERR LL_SPI_SR_CRCE
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: Can we avoid this as well?
You're not supposed to (re)define LL code inside zephyr

File: drivers/spi/spi_ll_stm32.h [hunk 19]
```diff
@@ -25,57 +25,126 @@ struct spi_stm32_data {
...
+		!LL_SPI_IsActiveFlag_TXC(spi));
+#else
 	return LL_SPI_IsActiveFlag_BSY(spi);
+#endif
 }
 
 static inline void ll_func_set_FIFO_threshold_one_data(SPI_TypeDef *spi)
 {
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+	LL_SPI_SetFIFOThreshold(spi, LL_SPI_FIFO_TH_01DATA);
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.h]: In case of other SPI FIFO IPs, threshold is set to 8-bit. Does 01data matches the same value?
In that case we can rename function as  `ll_func_set_fifo_threshold_8bit`

**Review by erwango — CHANGES_REQUESTED** [review]
last mile to walk..

File: drivers/spi/spi_ll_stm32.c (hunk 18 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: The goal was to "alias" the CRC error and TI frame error bits (which have the same purpose but a different name on this version of SPI peripheral)

We can also do this, but this chaining a bit more ifdefs : 

```
#ifdef CONFIG_SOC_SERIES_STM32MP1X

#define SPI_STM32_ERR_MSK (LL_SPI_SR_UDR | LL_SPI_SR_CRCE | LL_SPI_SR_MODF | \
			   LL_SPI_SR_OVR | LL_SPI_SR_TIFRE)

#else

#if defined(LL_SPI_SR_UDR)
#define SPI_STM32_ERR_MSK (LL_SPI_SR_UDR | LL_SPI_SR_CRCERR | LL_SPI_SR_MODF | \
			   LL_SPI_SR_OVR | LL_SPI_SR_FRE)
#elif defined(SPI_SR_FRE)
#define SPI_STM32_ERR_MSK (LL_SPI_SR_CRCERR | LL_SPI_SR_MODF | \
			   LL_SPI_SR_OVR | LL_SPI_SR_FRE)
#else
#define SPI_STM32_ERR_MSK (LL_SPI_SR_CRCERR | LL_SPI_SR_MODF | LL_SPI_SR_OVR)
#endif
#endif /* CONFIG_SOC_SERIES_STM32MP1X */
```

Comment (erwango) [reply drivers/spi/spi_ll_stm32.c]: @yboutreux this should be fine

**Comment by erwango** [comment]
Please rebase since dependency has been merged

**Comment by yboutreux** [comment]
> Please rebase since dependency has been merged

Change done and rebased, thank you !

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
Approved, but need to merge https://github.com/zephyrproject-rtos/hal_stm32/pull/18 first.

**Review by MaureenHelm — APPROVED** [review]
